onnxruntime/core/mlas/lib/q4_dq.cpp

```diff
@@ -638,6 +638,691 @@ struct BlockwiseQuantizer {
+     *        Since both src tensor and dst tensor are packed, it's not needed to consider sign
+     *        during the unpacking/packing in transpose.
+     * @param src_weights       The quantized weights, row major: [rows, columns] in qbits type.
+     *                          In uint8_t, size of [ceil(rows * columns * qbits / 8)].
+     * @param src_scales        [ceil(rows / quant_block_size), columns]
+     * @param src_zero_points   [ceil(rows / quant_block_size), columns] in qbits type. In uint8_t, size of
+     *                          [ceil(ceil(rows / quant_block_size) * columns * qbits / 8 )].
+     * @param dst_weights       the transposed quantized weights, column major. In uint8_t, the shape is
+     *                          [columns, ceil(rows / quant_block_size), ceil(quant_block_size * qbits / 8)]
+     * @param dst_scales        [columns, ceil(rows / quant_block_size)]
+     * @param dst_zero_points   [columns, ceil(ceil(rows / quant_block_size) * qbits / 8)] in uint8_t.
+     * @param rows              number of src rows in qbits type.
+     * @param columns           number of src columns in qbits type.
+     * @param quant_block_size  number of elements quantized together
+     * @param thread_pool       thread pool for parallel processing
+     */
+    static void TransposeColumnWiseQuantized(
+        const uint8_t* src_weights,
+        const Tin* src_scales,
+        const uint8_t* src_zero_points,
+        uint8_t* dst_weights,
+        Tin* dst_scales,
+        uint8_t* dst_zero_points,
+        int32_t rows,
+        int32_t columns,
+        int32_t quant_block_size,
+        MLAS_THREADPOOL* thread_pool
+    )
+    {
+        // Must avoid multiple thread write to a single byte, which means the starting index
+        // of a thread block must be even. To achieve that, we need to customize the thread
+        // block size based on the parity of columns.
+        if (columns & 1) {
+            TransposeColumnWiseQuantizedPackUnaligned(
+                src_weights, src_scales, src_zero_points,
+                dst_weights, dst_scales, dst_zero_points,
+                rows, columns, quant_block_size, thread_pool
+            );
+        } else {
+            TransposeColumnWiseQuantizedPackAligned(
+                src_weights, src_scales, src_zero_points,
+                dst_weights, dst_scales, dst_zero_points,
+                rows, columns, quant_block_size, thread_pool
+            );
+        }
+    }
+
+private:
+    static void QuantizeColumnWisePackAligned(
+        const Tin* src,
+        Tin* scales,
+        uint8_t* zero_points,
+        uint8_t* dst,
+        int32_t rows,
+        int32_t columns,
+        int32_t quant_block_size,
+        MLAS_THREADPOOL* thread_pool
+    )
+    {
+        ORT_ENFORCE(columns % 2 == 0, "Columns must be multiple of 2.");
+        // Thread block is [quant_block_size, thread_blk_size]. thread_blk_size % 2 == 0.
+        const int32_t thread_blk_size = 64;
```

Comment:
>64; [](http://example.com/codeflow?start=40&length=3)

64 is a little bit too small. Cache line size is 64 bytes on most CPUs, so it is better to be at least 128.

Reply:
updated to 128